framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -4631,6 +4613,442 @@ private void makeGroundTargetType(
+    if (capturedTypeMirror.getTypeArguments().isEmpty()) {
+      return false;
+    }
+
+    if (uncapturedType.wasRaw() || uncapturedType.containsUninferredTypeArguments()) {
+      return false;
+    }
+
+    if (capturedTypeMirror.getTypeArguments().size() != uncapturedType.getTypeArguments().size()) {
+      throw new BugInCF(
+          "Not the same number of type arguments: capturedTypeMirror: %s uncapturedType: %s",
+          capturedTypeMirror, uncapturedType);
+    }
+
+    for (int i = 0; i < capturedTypeMirror.getTypeArguments().size(); i++) {
+      TypeMirror capturedTypeArgTM = capturedTypeMirror.getTypeArguments().get(i);
+      AnnotatedTypeMirror uncapturedTypeArg = uncapturedType.getTypeArguments().get(i);
+      if (uncapturedTypeArg.getKind() != TypeKind.WILDCARD) {
+        continue;
+      }
+      if (TypesUtils.isCaptured(capturedTypeArgTM)
+          || capturedTypeArgTM.getKind() != TypeKind.WILDCARD) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Apply capture conversion.
+   *
+   * <p>Capture conversion is the process of converting wildcards in a generic type to fresh type
+   * variables. See <a
+   * href="https://docs.oracle.com/javase/specs/jls/se11/html/jls-5.html#jls-5.1.10">JLS 5.1.10</a>
+   * for all the details.
+   *
+   * <p>If {@code type} is not a declared type or if it does not have any wildcard type arguments,
+   * this method returns {@code type}.
+   *
+   * @param typeToCapture type to capture
+   * @return the captured type
+   */
+  public AnnotatedTypeMirror applyCaptureConversion(AnnotatedTypeMirror typeToCapture) {
+    TypeMirror capturedTypeMirror = types.capture(typeToCapture.getUnderlyingType());
+    return applyCaptureConversion(typeToCapture, capturedTypeMirror);
+  }
+
+  /**
+   * Apply capture conversion.
+   *
+   * <p>Capture conversion is the process of converting wildcards in a generic type to fresh type
+   * variables. See <a
+   * href="https://docs.oracle.com/javase/specs/jls/se11/html/jls-5.html#jls-5.1.10">JLS 5.1.10</a>
+   * for all the details.
+   *
+   * <p>If {@code type} is not a declared type or if it does not have any wildcard type arguments,
+   * this method returns {@code type}.
+   *
+   * @param type type to capture
+   * @param typeMirror type from the Java compiler that has captured wildcards, which are used as
+   *     the underlying type of the captured wildcards
+   * @return the captured type
+   */
+  public AnnotatedTypeMirror applyCaptureConversion(
+      AnnotatedTypeMirror type, TypeMirror typeMirror) {
+    if (!shouldCapture(type, typeMirror)) {
+      if (type.containsUninferredTypeArguments()
+          && typeMirror.getKind() == TypeKind.DECLARED
+          && type.getKind() == TypeKind.DECLARED) {
+        // If capture conversion is not applied because the type contains uninferred
+        // wildcards, then mark any uncaptured wildcards as "uninferred".
```

Comment:
The notion of uninferred wildcards is not in JLS 5.1.10.  Where is a description of the variant that this method computes?  Or are uninferred wildcards always replaced by inferred wildcards before this method terminates?  (In that case, the algorithm differs from 5.1.10 but the result does not.)

Reply:
Uninferred type arguments are a Checker Framework concept. They are a work around for 979.  See the comment on `AnnotatedTypeMirror#containsUninferredTypeArguments()`.